### PR TITLE
Add Cp utility 

### PR DESCRIPTION
**Pull request by wildcat26**

### CODE REVIEW

File: src/Streamly/Coreutils/Cp.hs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+{-# INLINE copy #-}
+copy
+    :: CpOptions
+    -> FilePath
+    -> FilePath
+    -> IO Bool
+copy opt src dst = do
+    existDst <- doesFileExist dst
+    existSrc <- doesFileExist src
+    if not existSrc
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:73]: if/else indentation

Comment (adithyaov) [reply src/Streamly/Coreutils/Cp.hs:73]: If the options are exclusive, use `when` instead.

File: src/Streamly/Coreutils/Cp.hs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+-- exists in the destination path and @overwriteExisting@ is set to
+-- @False@ or if the source file does not exist.
+--
+-- @since 0.1.0.0
+{-# INLINE copy #-}
+copy
+    :: CpOptions
+    -> FilePath
+    -> FilePath
+    -> IO Bool
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:69]: `copy :: CpOptions -> Path Abs File -> Path Abs File -> IO Bool`?

File: src/Streamly/Coreutils/Cp.hs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        File.fromChunks dst $ File.toChunksWithBufferOf (256*1024) src
+        return True
+
+
+-- | Append two paths - first one should be a directory
+-- Assumes second path is relative and without a leading '/'
+--
+-- @since 0.1.0.0
+{-# INLINE append #-}
+append
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:96]: You don't need this if you use `path`

File: src/Streamly/Coreutils/Cp.hs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+         else if end == pathSep
+         then dir ++ file
+         else dir ++ [pathSep] ++ file
+
+
+-- | Removes a trailing slash from a file path
+--
+-- @since 0.1.0.0
+{-# INLINE removeTrailingSlash #-}
+removeTrailingSlash
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:116]: You don't need this if you use `path`

File: src/Streamly/Coreutils/Cp.hs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+-- | Lists the contents (both files and directories) of the directory
+-- relative to the argument directory.
+-- @Left path@ represents that @path@ is a directory while @Right path@
+-- represents that it is a file.
+--
+-- @since 0.1.0.0
+traverseDir
+    :: IsStream t
+    => FilePath
+    -> t IO (Either FilePath FilePath)
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:139]: `traverseDir :: CpOptions -> Path Abs Dir -> t IO (Either (Path Abs Dir) (Path Abs File))`?

File: src/Streamly/Coreutils/Cp.hs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+-- to the destination directory.
+-- Arguments are of type @FilePath@.
+--
+-- @since 0.1.0.0
+copyDirToDir
+    :: IsStream t
+    => CpOptions
+    -> FilePath
+    -> FilePath
+    -> t IO ()
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:174]: `copyDirToDir :: CpOptions -> Path Abs Dir -> Path Abs Dir -> IO ()`?
Why `t IO ()`?

File: src/Streamly/Coreutils/Cp.hs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+-- | Copies the source file to the
+-- destination path with a possible rename.
+--
+-- @since 0.1.0.0
+cpFileWithRename
+    :: CpOptions
+    -> SomeBase File
+    -> SomeBase File
+    -> IO Bool
+cpFileWithRename opt src dst = do
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:212]: Why do you have this? Isn't this just `copy`?

File: src/Streamly/Coreutils/Cp.hs [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    copy opt srcFp dstFp
+
+
+-- | Copies the source file preserving the source file name
+-- in the destination directory.
+-- Returns @True@ in the @IO@ monad if copying was successful,
+-- @False@ otherwise.
+--
+-- @since 0.1.0.0
+cpFile
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:224]: Why do you have this? Why not just use `Path Abs File` in `copy`?

File: src/Streamly/Coreutils/Cp.hs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+            Rel fpath -> fromRelFile $ filename fpath
+
+
+-- | Copies a stream of files to the destination directory.
+-- The stream can be drained with the appropriate combinator
+-- to achieve concurrency.
+-- Fails if the destination directory does not exist.
+--
+-- @since 0.1.0.0
+cpFiles
```
Comment (adithyaov) [inline src/Streamly/Coreutils/Cp.hs:253]: What is `SomeBase`? Why not use `Path`?

File: src/Streamly/Coreutils/Cp.hs (hunk 6 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:174]: I did not drain the stream so that user can use the appropriate combinator like `asyncly`, `wAsyncly` to achieve concurrency.

File: src/Streamly/Coreutils/Cp.hs (hunk 7 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:212]: yes, just that `copy` isn't exposed. This is.

File: src/Streamly/Coreutils/Cp.hs (hunk 8 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:224]: This is the `File -> Dir` case.

File: src/Streamly/Coreutils/Cp.hs (hunk 9 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:253]: http://hackage.haskell.org/package/path-0.8.0/docs/Path-Posix.html#t:SomeBase

File: src/Streamly/Coreutils/Cp.hs (hunk 5 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs]: :+1: `CpOptions` isn't needed while traversing a directory.

File: src/Streamly/Coreutils/Cp.hs (hunk 1 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:75]: https://github.com/composewell/haskell-dev/blob/master/coding-style.rst#if-then-else
What is the issue with this indentation style?

File: src/Streamly/Coreutils/Cp.hs [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+-- @Left path@ represents that @path@ is a directory while @Right path@
+-- represents that it is a file.
+--
+-- @since 0.1.0.0
+traverseDir
+    :: IsStream t
+    => Path Abs Dir
+    -> t IO (Either (Path Abs Dir) (Path Abs File))
+traverseDir baseDir = do
+    IP.concatMapTreeWith ahead listContents $ S.yield $ Left baseDir
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:104]: We should be able to choose either to run this in serial or ahead style, instead of ahead always

File: src/Streamly/Coreutils/Cp.hs [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        else do
+            actualPath <- stripAndPrepend dir src dest
+            createDirectoryIfMissing True $ fromAbsDir actualPath
+    handleForce (Right file) = do
+        actualPath <- stripAndPrepend file src dest
+        _ <- copy opt file actualPath
+        return ()
+
+
+    handleNonForce :: Either (Path Abs Dir) (Path Abs File) -> IO Bool
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:164]: Why do you have two identical functions `handleForce` and `handleNonForce`?

File: src/Streamly/Coreutils/Cp.hs [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    :: IsStream t
+    => CpOptions
+    -> t IO (Path Abs File)
+    -> Path Abs Dir
+    -> t IO (Path Abs File)
+    -- ^ Stream of files for which copying was successful
+cpFiles opt strm dir = do
+    if force opt == True
+    then
+        S.filterM (\f -> cpFile opt f dir) strm
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:217]: Why is one of the case using `filterM` and the other `takeWhileM` and how is this related to `force` option.

File: src/Streamly/Coreutils/Cp.hs [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+       isFile <- doesFileExist $ fromAbsFile filePath
+       if isFile == True
+       then do
+           return $ Right filePath
+       else do
+           asDir <- parseRelDir path
+           return $ Left $ dir </> asDir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory dir
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:125]: No need to have an `unsafePerformIO`, you can change `listContents` appropriately to use `contents :: FilePath -> IO [FilePath]`

File: src/Streamly/Coreutils/Cp.hs [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+        putStrLn $ "overwriteExisting : False " ++ dstFp ++ " already exists"
+        return False
+    else if verbose opt == True
+    then do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        putStrLn $ srcFp ++ " -> " ++ dstFp
+        return True
+    else do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        return True
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:90]: Doesn't look like you're handling the appropriate use case when `force` is true and you don't have the required permission.

File: src/Streamly/Coreutils/Cp.hs [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+-- | Prints paths of source, destination files when @verbose@ is @True@.
+-- Returns @True@ if copying operation was successful.
+-- @False@ is returned in cases where a file by the same name
+-- exists in the destination path and @overwriteExisting@ is set to
+-- @False@ or if the source file does not exist.
+--
+-- @since 0.1.0.0
+{-# INLINE copy #-}
+copy
```
Comment (pranaysashank) [inline src/Streamly/Coreutils/Cp.hs:65]: This function`copy` is being used to map and copy files in a stream, if this function results in an IOError either due to not having the necessary permissions or failure to create a file the entire program crashes. But I think the correct thing to do here is to notify the error message to the user and continue copying rest of the files.

File: streamly-coreutils.cabal [hunk 16]
```diff
@@ -70,10 +76,12 @@ benchmark coreutils-bench
 test-suite coreutils-test
   import: compile-options
   type: exitcode-stdio-1.0
-  main-is: Main.hs
-  hs-source-dirs: test
+  main-is: Cp.hs
```
Comment (pranaysashank) [inline streamly-coreutils.cabal:79]: Did you forget to push this change?

File: src/Streamly/Coreutils/Cp.hs (hunk 15 above)
Comment (wildcat26) [reply src/Streamly/Coreutils/Cp.hs:65]: depends on whether the `force` flag is true or not. If it is, then we should continue, else not.

Comment (pranaysashank) [reply src/Streamly/Coreutils/Cp.hs:65]:
```bash
[pranaysashank issue]$ ls -l
total 0
---------- 1 pranaysashank pranaysashank 0 Aug 17 12:39 A
-rw-r--r-- 1 pranaysashank pranaysashank 0 Aug 17 12:39 B
-rw-r--r-- 1 pranaysashank pranaysashank 0 Aug 17 12:39 C
[pranaysashank issue]$ mkdir dest
[pranaysashank issue]$ cp B A C dest/
cp: cannot open 'A' for reading: Permission denied
[pranaysashank issue]$ ls dest/
B  C
```
Look at this example on my machine, it failed to copy A because it doesn't have the permission to read, but it continued copying the remaining files i.e. C
